src/contractors.jl

```diff
@@ -0,0 +1,125 @@
+# contractors:
+"""
+    Contractor{F}
+
+    Abstract type for contractors.
+"""
+abstract type Contractor{F} end
+
+export Bisection, Newton
+
+# bisection
+struct Bisection{F} <: Contractor{F}
+    f::F
+end
+
+function (contractor::Bisection)(X, tol)
+    image = contractor.f(X)
+
+    if !(contains_zero(image))
+        return :empty, X
+    end
+
+    return :unknown, X
+end
+
+# Newton
+struct Newton{F,FP,O} <: Contractor{F}
+    f::F
+    f_prime::FP
+    op::O
+end
+
+function Newton(f::Function, f_prime::Function)
+    Newton(f, f_prime, 𝒩)
+end
+
+function (C::Newton)(X, tol)
+    # use Bisection contractor for this:
+    if !(contains_zero(IntervalBox(C.f(X))))
+        return :empty, X
+    end
+
+    # given that have the Jacobian, can also do mean value form
+
+    NX = C.op(C.f, C.f_prime, X) ∩ X
+
+    isempty(NX) && return :empty, X
+
+    if NX ⪽ X  # isinterior; know there's a unique root inside
+        NX =  refine(X -> C.op(C.f, C.f_prime, X), NX, tol)
+        return :unique, NX
+    end
+
+    return :unknown, NX
+end
+
+
+# What is this guarded_mid for? Shouldn't it be checking if f(m)==0?
+doc"""Returns the midpoint of the interval x, slightly shifted in case
+the midpoint is an exact root"""
+function guarded_mid{T}(f, x::Interval{T})
+    m = mid(x)
+
+    if f(m) == 0                      # midpoint exactly a root
+        α = convert(T, 0.46875)      # close to 0.5, but exactly representable as a floating point
+        m = α*x.lo + (one(T)-α)*x.hi   # displace to another point in the interval
+    end
+
+    @assert m ∈ x
+
+    m
+end
+
+doc"""
+Single-variable Newton operator
+"""
+function 𝒩{T}(f, x::Interval{T}, deriv::Interval{T})
```

Comment:
I think this method does exactly what the one in lines 89-93 does. This method is used in `newton_refine`, which I understand will be replaced by `refine`, which uses the method of lines 89-93. If this is so, I suggest to delete it.

Reply:
I left the old stuff in for the moment so that we can continue to use `newton` etc. for the moment, e.g. for performance comparisons.